Add SES lockdown to extension webapp

EtDu opened the pull request:
Adds `lockdown()` to background, UI, contentscript & phishing-detect (all contexts except in-page)

### SES + Sentry integration
Sentry integrates itself by overriding some intrinsics like `Function.prototype.toString` and platform apis like `console`. We carefully install only Sentry before performing the lockdown. Then start the app normally.

### SES compat via `patch-package`
- `symbol-observable` modifies `Symbol` intrinsic (pattern used by multiple packages, in our case Redux). Once we get these upstream changes merged, we can remove the patch
  - PR made in `redux` (merged, awaiting publish)
  - PR made in `redux-devtools-instrument` (merged, updated in `remote-redux-devtools`)
  - PR made in `reactiveX` (not yet merged)

### Summary of Changes:
- more mem for box (medium+)
- more "max old mem size" to 2gb
- add `patch-package` for workaround with `symbol-observable@1`
- run sentry and lockdown before app init
- new bundle for setting up sentry
- replace `envify` with `loose-envify` for parsing error workaround
- add METAMASK_VERSION env var
- pass down `process.env` to child processes
- log error code if child process fails
- remove `lavamoat-core`, use `ses`

EtDu committed:
Freezeglobals: remove Promise freezing, add lockdown

EtDu committed:
background & UI: temp disable sentry

EtDu committed:
add loose-envify, dedupe symbol-observable

EtDu committed:
use loose envify

EtDu committed:
add symbol-observable patch

EtDu committed:
run freezeGlobals after sentry init

EtDu committed:
use require instead of import

EtDu committed:
add lockdown to contentscript

EtDu committed:
add error code in message

EtDu committed:
try increasing node env heap size to 2048

EtDu committed:
change back circe CI option

EtDu committed:
make freezeGlobals an exported function

EtDu committed:
make freezeGlobals an exported function

EtDu committed:
use freezeIntrinsics

EtDu committed:
pass down env to child process

EtDu committed:
fix unknown module

EtDu committed:
fix tests

EtDu committed:
change back to 2048

EtDu committed:
fix import error

EtDu committed:
attempt to fix memory error

EtDu committed:
fix lint

EtDu committed:
fix lint

EtDu committed:
fix mem gain

EtDu committed:
use lockdown in phishing detect

EtDu committed:
fix lint

EtDu committed:
move sentry init into freezeIntrinsics to run lockdown before other imports

EtDu committed:
lint fix

kumavis reviewed (commented):
I think this will work for our purposes but is not valid esm (running code between import statements). To make it work we should run the lockdown as part of the module initialization. If we don't have a great way to add the context (ui, etc) to the lockdown we can create lockdowns for each context that hard code it 😐

EtDu commented:
/sparkle

EtDu committed:
custom lockdown modules per context

EtDu committed:
lint fix

EtDu committed:
fix global test

EtDu committed:
remove run in child process

EtDu committed:
remove lavamoat-core, use ses, require lockdown directly

EtDu committed:
revert childprocess

EtDu committed:
patch package postinstall

EtDu committed:
revert back child process

EtDu committed:
add postinstall to ci

EtDu committed:
revert node max space size to 1024

EtDu committed:
put back loose-envify

EtDu committed:
Disable sentry to see if e2e tetss pass

EtDu committed:
use runLockdown, add as script in manifest

EtDu committed:
remove global and require from runlockdown

EtDu committed:
add more memory to tests

EtDu committed:
upgrade resource class for prep-build & prep-build-test

EtDu committed:
fix lint

EtDu committed:
lint fix

EtDu committed:
upgrade remote-redux-devtools

EtDu committed:
skillfully re-add sentry

EtDu committed:
lintfix

EtDu committed:
fix lint

EtDu committed:
put back beep

Gudahtt commented on app/scripts/ui.js:
```diff
@@ -16,28 +12,21 @@ import EthQuery from 'eth-query'
   ENVIRONMENT_TYPE_POPUP,
 } from './lib/enums'
 import { getEnvironmentType } from './lib/util'
+/* eslint-disable import/first,import/order */
```
Is this still needed? I can't tell why it was added

EtDu replied:
removed!

Gudahtt commented on package.json:
```diff
@@ -49,7 +49,8 @@
     "update-changelog": "./development/auto-changelog.sh",
-    "generate:migration": "./development/generate-migration.sh"
+    "generate:migration": "./development/generate-migration.sh",
+    "postinstall": "npx patch-package"
```
This seems unsafe. If we're going to use `patch-package`, it should at least be included as a dev dep and included in the lockfile, so we're not downloading a new `@latest` version each time.

Gudahtt commented on package.json:
```diff
@@ -130,6 +131,7 @@
     "jsonschema": "^1.2.4",
     "lodash": "^4.17.19",
     "loglevel": "^1.4.1",
+    "loose-envify": "^1.4.0",
```
This should be under `devDependencies`, since it's used by the build system. Also, shouldn't this replace `envify`? I noticed that `envify` is still listed

Gudahtt commented on development/build/static.js:
```diff
@@ -44,6 +44,16 @@ const copyTargets = [
     dest: ``,
   },
+  {
+    src: `./node_modules/ses/dist/`,
```
Would it be possible to import this from `runLockdown` instead? It would be nice to avoid copying this directly from the dependency. Aside from being unconventional, this this is also incompatible with package managers that don't use `node_modules` (e.g. Yarn v2).

kumavis replied:
@Gudahtt  part of the reasoning here is to take the ses dist as is without any transformations. also possible under yarn@2, would just need to rework the import

Gudahtt commented on app/scripts/initSentry.js:
```diff
@@ -0,0 +1,7 @@
+
+// setup sentry error reporting
+global.sentry = setupSentry({
+  release: process.env.METAMASK_VERSION,
```
I'm guessing that `METAMASK_VERSION` was added so it could be used here?

We use `platform.getVersion()` elsewhere, which calls `extension.runtime.getManifest().version`. This should work here as well I think, since this API is supported in both the background and the contenscript.

kumavis replied:
correct. we wanted to avoid importing anything e.g. the extensionizer package during this limited setup phase

Gudahtt commented on app/scripts/runLockdown.js:
```diff
@@ -0,0 +1,7 @@
+// Freezes all intrinsics
+// eslint-disable-next-line no-undef,import/unambiguous
+lockdown({
+  errorTaming: 'unsafe',
```
So by setting these three to "unsafe", are they completely unaffected by the lockdown?

I tried looking this up in the SES docs, but I couldn't find where this was documented.

I was going to ask about whether Errors still had stack traces now, as I had read that SES disabled stack traces for security reasons, and had some alternate way to get them. But I'm guessing that this leaves them exactly as-is for now, which is good.

kumavis replied:
this part of SES has changed recently and so i would not consider it stable, but for now:
- the `errorTaming: 'unsafe'` allows you to retain error stacks. error stacks are normally hidden so that code cant observe the context in which its called. ses has a feature where the error stack is hidden but can be recovered for eg logging to console. this feature is not exposed yet so we cant use it to expose to sentry.
- all other `*Taming: 'unsafe'` allow potentially powerful/dangerous but expected things like non-determinism via `Math.random` and `new Date()`
- we leave security on for `RegExp.lastMatch` as it is not used much in practice

kumavis replied:
more fine grained control over access to features like Math.random will come with LavaMoat 🌋

EtDu replied:
That is correct. Setting `errorTaming` to `unsafe` reveals stack traces. Math is for `math.random` & Date is for `Date.now`. They aren't unaffected by lockdown, but instead accessible via the global namespace. SES disables access by default for security concerns.

Gudahtt commented on .circleci/config.yml:
```diff
@@ -122,8 +122,9 @@ jobs:
   prep-build:
     docker:
       - image: circleci/node@sha256:e16740707de2ebed45c05d507f33ef204902349c7356d720610b5ec6a35d3d88
+    resource_class: medium+
```
What was the reason for increasing this? Was this just to make the intermittent failures less likely, or did something introduced in this PR affect these jobs?

kumavis replied:
this helped alleviate CI failures. I think we run 3x parallel node processes with 2gig "max old space" => `medium+` for 6GB mem

kumavis replied:
we could try removing this now that we're no longer parsing and transforming ses lockdown dist

EtDu replied:
`prep-build` & `prep-build-test` was consistently failing with out of memory. These steps from other PRs seemed to be passing just fine, so it must have been something from here that caused it. I'm not sure what though.

Gudahtt reviewed (commented):
That seems like a clever way to keep Sentry working! Though I'm surprised this worked - I would have expected the lockdown to "undo" global mutations as well as preventing new ones. Though I don't know - maybe it doesn't work like that. We can test this either way with real Sentry events, targeting our test Sentry project.

I wonder if there is some way to operate Sentry in a reduced capacity as well, such that global mutations are not necessary :thinking:

EtDu committed:
remove envify, add loose-envify and patch-package in dev deps

kumavis commented:
CI is listing a pending check (test-deps) but it in fact passed

Gudahtt commented:
Since there is a working upstream version of `symbol-observable`, we can use a Yarn resolution to update transitive dependencies instead of a patch.

e.g. in the `resolutions` field of `package.json`:
```
    "**/symbol-observable": "^2.0.3",
```

We still need `patch-package` in cases where there is no upstream solution at all, but I think using resolutions to force-update transitive deps is a less fragile approach than patching.

Gudahtt commented:
I have tested this locally, and have confirmed that Sentry still works correctly! The error was rate-limited so I didn't see it in the dashboard, but the network request looked correct to me.

Gudahtt committed:
Replace patch with Yarn resolution (#9923)

Instead of patching `symbol-observable`, this ensures that all
versions of `symbol-observable` are resolved to the given range, even
if it contradicts the requested range.
